### PR TITLE
Document the meaning of Option in Contract#agreementText

### PR DESCRIPTION
**Pull request by leo-da**
Documenting contract class.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/docs/source/support/release-notes.rst), if appropriate

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by S11001001** [comment]
Perhaps instead we should return `None` for the empty string case instead of `Some("")` instead, then?

**Comment by leo-da** [comment]
I thought about this... I was told new contracts may have agreement set to an empty string in DAML. This will be a contract with agreement field set to "", which we will have to decode to None, which is wrong... the field is present. So, I think empty string is all right. If you insist on converting it to None, I am fine with that, it is still better than having `Some("")`

**Comment by leo-da** [comment]
```
template EmptyAgreement
  with
    owner: Party
  where
    signatory owner
    agreement ""
```

**Comment by hurryabit** [comment]
To my understanding the agreement text is mandatory in DAML-LF. You can omit in DAML though, but then we fill in the empty string for you, see https://github.com/digital-asset/daml/blob/8ec03875c682b016082bb2ca83eb4f5c00fa6fae/daml-foundations/daml-ghc/daml-stdlib-src/DA/Internal/Template.daml#L110.

@bitonic Can you please confirm the statement regarding DAML-LF.

**Comment by leo-da** [comment]
@hurryabit this is coming from the ledger event. We have to support older versions of the ledger api. Agreement was recently added to the ledger api. In one week nobody will be using this that old version of the ledger api, but we still have to support interoperability with older clients. I guess this is more of an exercise than a real use case. I am also not sure why it is so important to get `None` for old client. I just give up at this point. But `None` currently means that we don't know if the `agreement` was specified in DAML or not.

**Comment by bitonic** [comment]
@hurryabit what @leo-da said -- we need to be able to use new client with old servers.

**Comment by hurryabit** [comment]
@leo-da `None` can't mean that you don't know if the agreement was specified in DAML because no agreement clause at all and `agreement ""` are not distinguishable in DAML (and hence DAML-LF). I suppose `None` means the server hasn't sent the agreement text instead.

**Comment by bitonic** [comment]
@hurryabit @leo-da I did not read the code, but on the API we wrap the agreement text in `StringValue` precisely to distinguish empty strings from absent strings. Bindings should provide an `Optional[String]` view, where `None` means "I'm talking to a server that does not support agreement text, and `Some(txt)` means "I'm talking to a server that supports agreement texts".

File: language-support/scala/bindings/src/main/scala/com/digitalasset/ledger/client/binding/binding/Contract.scala [hunk 1]
```diff
@@ -5,6 +5,17 @@ package com.digitalasset.ledger.client.binding
 
 import com.digitalasset.ledger.api.v1.{value => rpcvalue}
 
+/**
+  * A class representing a DAML contract of specific type (DAML template) with assigned contract ID and agreement text.
+  *
+  * @param contractId     Contract ID.
+  * @param value          Contract instance as defined in DAML template (without `contractId` and `agreementText`).
+  * @param agreementText  Agreement text. `None` means we do not know if the `agreementText` was specified on the contract.
```
Comment (bitonic) [inline language-support/scala/bindings/src/main/scala/com/digitalasset/ledger/client/binding/binding/Contract.scala]:
```suggestion
  * @param agreementText  Agreement text. `None` means that we did not receive the `agreementText` from the server.
```

**Comment by leo-da** [comment]
@hurryabit that confused me a lot... but yeah technically we can get a `None` but we probably will never receive it in a week once everyone is using the ledger-api that supports `agreementText` field.

ledger-api and daml-lf are two different specs. I am looking forward to see how we will support interoperobility between different ledger-api clients, given:
- ledger-api does not specify ledger-api version in the spec and
- we don't have anti-corruption layer (ADTs) sitting between protobuf messages and business logic.

**Pull request merged**